src/nupic/engine/RegionImpl.hpp

```diff
@@ -129,14 +132,33 @@ class RegionImpl
   virtual std::string executeCommand(const std::vector<std::string> &args,
                                      Int64 index) = 0;
 
-  // Per-node size (in elements) of the given output.
-  // For per-region outputs, it is the total element count.
-  // This method is called only for outputs whose size is not
-  // specified in the nodespec.
-  virtual size_t getNodeOutputElementCount(const std::string &outputName) = 0;
-
   /* -------- Methods that may be overridden by subclasses -------- */
 
+  // Buffer size (in elements) of the given input/output.
+  // It is the total element count.
+  // This method is called only for buffers whose size is not
+  // specified in the nodespec.  This is used to allocate
+  // buffers during initialization.
+  // Return 0 for outputs that are not used or size does not matter.
+  virtual size_t getNodeInputElementCount(const std::string &outputName) const {
+    return 0;
+  }
+  virtual size_t getNodeOutputElementCount(const std::string &outputName) const {
```

Comment:
> are these methods needed?

Yes, they can be overridden by the Region implementations. (thats why they are virtual).
Yes, they are needed because they are part of the API.  Region implementations are using it.
The change I made was that they are no longer REQUIRED to override because
The function they perform is also covered by the new function askImplForOutputDimensions( ) which returns a full dimension.  New region implementations should override this new function.

I will add some more documentation about this.  Obviously it was not clear enough.